### PR TITLE
CAN-357: Refactor to use real data ingestion

### PR DESCRIPTION
**Pull request by anders0204**
- Refactored the visualisation tool to enable use of real data with the decipher data handler package.
- Also removed the Person class and everything related to predictions.
- Functionality that _might_ be used in the future, such as vaccination data and the plots with predictions is left as comments (mostly)
- Added `data_ingestion.py` which includes transformers and pipelines to prepare the data into a format that the rest of the application (which is largely untouched) can deal with
- The `NAHandler` transformers is required only due to the processing from the data handler, which gives very annoying `<NA>` (`pd._libs.missing.NA`) types. I convert them to good old `np.nan` instead.

The demo data is now the random mock data that we have been using for testing, so the plots look weird. Can you see if they are weird because this or if there's a bug somewhere?

### CODE REVIEW

File: bokeh-demo/bokeh_demo/backend.py [hunk 1]
```diff
@@ -310,7 +155,7 @@ def select_person_callback(attr, old, selected_people):
     def set_group_selected_callback(attr, old, new):
         if new == []:  # Avoid unsetting when hitting a line in scatter plot
             return
-        selected_people = list({exam_source.data["person_index"][i] for i in new})
+        selected_people = list({exam_source.data["PID"][i] for i in new})
```
[the remark this answers is not in the document]
Comment (Northo) [reply bokeh-demo/bokeh_demo/backend.py:158]: After looking even more, I think the issue has to do with the exam indices. They are now not sequential.

File: bokeh-demo/bokeh_demo/data_ingestion.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    def fit(self, X: pd.DataFrame, y=None) -> CategoryColumnConverter:
+        self.category_cols = X.select_dtypes(include=["category"]).columns
+        return self
+
+    def transform(self, X: pd.DataFrame) -> pd.DataFrame:
+        X = X.copy()
+        for col in self.category_cols:
+            X[col] = X[col].apply(lambda x: x.value)
+        X[self.category_cols] = X[self.category_cols].astype("str")
+        return X
```
[the remark this answers is not in the document]
Comment (Northo) [reply bokeh-demo/bokeh_demo/data_ingestion.py:45]: This needs adressing, due to `DataManager` not giving enums back for parquet.

```python
    @staticmethod
    def try_convert_enum(x):
        try:
            return x.value
        except AttributeError:
            return x
....

      X[col] = X[col].apply(self.try_convert_enum)
```
seems to do the trick.

File: bokeh-demo/bokeh_demo/backend.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Northo) [reply bokeh-demo/bokeh_demo/backend.py:158]: Long rant here.... I'll continue to investigate, but release the PR for now

[the remark this answers is not in the document]
Comment (Northo) [reply bokeh-demo/bokeh_demo/backend.py:158]: I played a bit with keeping it mostly as-is, but changing 
```python
        all_indices = [
            exam_index
            for exam_index, person_index in zip(exam_source.data["index"], exam_source.data["PID"])
            if person_index in selected_people
        ]
```
I think that is on to something, but don't have the time to investigate. We have to know if the person_source.selected.indices expects the index of the list, or the index according to the 'index' field.

**Comment by Northo** [comment]
Great work! massive PR
My suggestions
- [ ] Use parquet loading instead
- [ ] See comments on the various transformers
- [ ] Remove `matfact` as dependency, together with the examples dependency group

There is something wrong in `bokeh_demo.backend.link_sources`. I have not been able to find out exactly what, but I believe it has to to with the following: before, the indices of both person_source and exams_source was sequential. Now, at least ofr exam_source, the index is not sequential. Might be that simply resetting the index will help...

**Review by Northo — CHANGES_REQUESTED** [review]
Great work! massive PR
My suggestions
- [ ] Use parquet loading instead
- [x] See comments on the various transformers
- [x] Remove `matfact` as dependency, together with the examples dependency group

There is something wrong in `bokeh_demo.backend.link_sources`. I have not been able to find out exactly what, but I believe it has to to with the following: before, the indices of both person_source and exams_source was sequential. Now, at least ofr exam_source, the index is not sequential. Might be that simply resetting the index will help...

File: bokeh-demo/bokeh_demo/data_ingestion.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    def fit(self, X: pd.DataFrame, y=None) -> NAHandler:
+        self.bool_cols += X.select_dtypes(include=["bool"]).columns.to_list()
+        return self
+
+    def transform(self, X: pd.DataFrame) -> pd.DataFrame:
+        X = X.copy()
+        if self.bool_cols:
+            X[self.bool_cols] = X[self.bool_cols].astype("float")
+        X = X.fillna(np.nan)
+        return X
```
[the remark this answers is not in the document]
Comment (anders0204) [reply bokeh-demo/bokeh_demo/data_ingestion.py:26]: No but like we discussed, line 25 is redundant so I will just remove that one?

File: bokeh-demo/bokeh_demo/frontend.py [hunk 4]
```diff
@@ -157,14 +157,14 @@ def __init__(self, source_manager: SourceManager):
             color={
                 "expr": CustomJSExpr(
                     args={"colors": self._marker_colors},
-                    code=f"return this.data.{self._marker_color_key}.map(i => colors[i]);",  # noqa: E501
+                    code=f"return Array.from(this.data.{self._marker_color_key}).map(i => colors[i]);",  # noqa: E501
```
[the remark this answers is not in the document]
Comment (anders0204) [reply bokeh-demo/bokeh_demo/frontend.py:160]: Haha [ask yourself](https://github.com/Simula-Consulting/decipher/commit/3ba65b1611ee225a69707881b13f69329ddaff8b#diff-35620526ef2e5c752cf1ec9c15f1a416dcaa2a7be3e31024ecc2a9f69ba8939d)!

File: bokeh-demo/bokeh_demo/data_ingestion.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+        # Adding dummy vaccine data for future use
+        person_df[["vaccine_age", "vaccine_year", "vaccine_type"]] = None
+        person_df[["vaccine_line_endpoints_age", "vaccine_line_endpoints_year"]] = [
+            ([], []) for _ in range(person_df.shape[0])
+        ]
+        return person_df
+
+    def _minmax(self, column):
+        return (min(column), max(column))
```
[the remark this answers is not in the document]
Comment (anders0204) [reply bokeh-demo/bokeh_demo/data_ingestion.py:101]: staticmethod ✅ , the transform method is already to cluttered I feel like

File: bokeh-demo/bokeh_demo/data_ingestion.py [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+            pid: sub_df["risk"].to_list() for pid, sub_df in X.groupby("PID")
+        }
+        self.person_exam_ages_map = {
+            pid: (sub_df["age"] / pd.Timedelta(days=365)).to_list()
+            for pid, sub_df in X.groupby("PID")
+        }
+        return self
+
+    def transform(self, X: pd.DataFrame) -> pd.DataFrame:
+        # Generate the person dataframe based on the exams dataframe
```
[the remark this answers is not in the document]
Comment (anders0204) [reply bokeh-demo/bokeh_demo/data_ingestion.py:65]: That was not the plan!

File: bokeh-demo/bokeh_demo/data_ingestion.py [hunk 7]
```diff
@@ -40,13 +38,22 @@ def fit(self, X: pd.DataFrame, y=None) -> CategoryColumnConverter:
...
+            return x
+
 
 class CreatePersonSource(BaseEstimator, TransformerMixin):
-    """Takes in the exams dataframe and returns another dataframe for plotting Lexis plots."""
+    """Takes in the exams dataframe and returns another dataframe (with one row per person)
+    for plotting Lexis plots."""
 
     person_results_map: dict[int, list[int]] = {}
     person_exam_ages_map: dict[int, list[int]] = {}
```
Comment (Northo) [inline bokeh-demo/bokeh_demo/data_ingestion.py:59]: Declare and describe this. Neat idea to simply use a lookup table!

```suggestion
    person_exam_ages_map: dict[int, list[int]] = {}
    person_exam_inds_map: dict[int, list[int]]
    """Mapping from PID to list of that persons exams' indices"""
```

File: bokeh-demo/examples/pilot/main.py [hunk 8]
```diff
@@ -199,26 +145,33 @@ class HomePlaces(str, Enum):
...
-    source_manager = SourceManager.from_people(people)
+    data_manager = DataManager.read_from_csv(
+        settings.data_paths.screening_data_path, settings.data_paths.dob_data_path
+    )
+    exams_df = data_manager.exams_df
+
+    exams_df = extract_people_from_pids([], exams_df)
+
+    exams_df = exams_pipeline.fit_transform(exams_df)
+    person_df = CreatePersonSource().fit_transform(exams_df)
```
[the remark this answers is not in the document]
Comment (Northo) [reply bokeh-demo/examples/pilot/main.py:177]: I suppose I can let it go..

[the remark this answers is not in the document]
Comment (Northo) [reply bokeh-demo/examples/pilot/main.py:177]: But suggest adding a comment.
```suggestion
    # Warning!
    # This is not the same DataFrame as `DataManager.person_df`.
    person_df = CreatePersonSource().fit_transform(exams_df)
```

File: bokeh-demo/pyproject.toml [hunk 9]
```diff
@@ -10,6 +10,7 @@ packages = [{include = "bokeh_demo"}]
 python = ">=3.10,<3.11"
 bokeh = "^3.0.2"
 pydantic = "^1.10.2"
+decipher = {git = "https://github.com/Simula-Consulting/decipher_data_handler"}
```
[the remark this answers is not in the document]
Comment (Northo) [reply bokeh-demo/pyproject.toml:13]: Will instead simply tag on github and use that.

[the remark this answers is not in the document]
Comment (Northo) [reply bokeh-demo/pyproject.toml:13]:
```suggestion
decipher = { git = "https://github.com/Simula-Consulting/decipher_data_handler", extras = ["pyarrow"], tag = "0.1.32"}
```

File: processing/tests/test_datasets/metadata.json [hunk 10]
```diff
@@ -0,0 +1 @@
+{"decipher_version": "0.1.29"}
```
Comment (Northo) [inline processing/tests/test_datasets/metadata.json:1]: If we update as written above, remember to include updating the version tag here.

File: bokeh-demo/pyproject.toml (hunk 9 above)
[the remark this answers is not in the document]
Comment (Northo) [reply bokeh-demo/pyproject.toml:13]: BTW. version 0.1.32 was set simply because that is the newest, no issue setting 0.1.29 instead, which I believe is what you tested against.

**Review by Northo — APPROVED** [review]
Well done!

See minor inline comments.

Found the following "functional" issues:
- [ ] Exams with nan/non-1-4 risk are not shown, but they are still there "invisible" but clickable. If we want them shown, we must add a color map for them.
- [ ] Clicking twice on a life line makes all exams selected, but not all life lines
- [ ] Y-axis on age/person lexis plot is the PID, not a person index based on range. This means people are super far from each other...
image

I believe all of these are fine to simply create issues for, and deal with later.

File: processing/tests/test_datasets/metadata.json (hunk 10 above)
Comment (anders0204) [reply processing/tests/test_datasets/metadata.json:1]: Temporarily set the version tag to 0.1.29 to make sure it's working with the rest in this branch. This is more up to date on [CAN-358/add/filters-on-landing-page](https://github.com/Simula-Consulting/decipher/tree/CAN-358/add/filters-on-landing-page)

Comment (anders0204) [reply processing/tests/test_datasets/metadata.json:1]: It now crashes if you try to serve without having pip installed pyarrow manually, but whatever, it's fixed on the new branch
